### PR TITLE
[terraform-resources] more robust vault secrets update

### PR DESCRIPTION
**Pull request by chassing**
Make the update of secrets in Vault more robust. 

The `disable_keys` method could raise an error, and the Vault secret updates are lost. Especially if the integration takes longer, e.g., creating an MSK cluster, the cached bundle version is outdated, and `disable_keys` raises a 404 error.

Write the outputs to Vault as soon as possible (and retry it) to avoid such cases.

### CODE REVIEW

**Review by janboll — APPROVED** [review]

File: reconcile/terraform_resources.py [hunk 1]
```diff
@@ -428,6 +432,11 @@ def run(
             dry_run, oc_map, ri, thread_pool_size, caller=acc_name
         )
 
+    # update secrets in vault immediately after applying them to the clusters
+    # so that nothing can raise an error before the secrets are updated
+    if actions and vault_output_path:
+        write_outputs_to_vault(vault_output_path, resource_specs)
```
Comment (maorfr) [inline reconcile/terraform_resources.py:438]: i would use `defer` here to avoid so that nothing can raise an error before disabling keys

Comment (hemslo) [reply reconcile/terraform_resources.py:438]: Here we only moved `write_outputs_to_vault` up, how about other steps after `disable_keys`, those cleanup steps can be skipped too.

Comment (chassing) [reply reconcile/terraform_resources.py:438]: As @hemslo said, it's about making `write_outputs_to_vault` more bulletproof. In the worst, worst, worst case, `disabled_keys` would run at the next `terraform-resources` reconciliation loop but it runs in `aws-iam-keys` anyways. It also doesn't depend on the current run, but `write_outputs_to_vault` does.

@hemslo, there isn't anything after `disabled_keys`

Comment (maorfr) [reply reconcile/terraform_resources.py:438]: the runs of aws-iam-keys is different if it runs from terraform-resources.

Comment (chassing) [reply reconcile/terraform_resources.py:438]: Ah, I see 👍 , but anyway, a missed `write_outputs_to_vault` is worse than a missed `disabled_keys`.

Comment (maorfr) [reply reconcile/terraform_resources.py:438]: if only we had something that will allow us not to miss anything, like `defer`.

Comment (hemslo) [reply reconcile/terraform_resources.py:438]: if `disabled_keys` throw error, then `cleanup_and_exit` won't run

Comment (hemslo) [reply reconcile/terraform_resources.py:438]: if we expect `disable_keys` to throw error, how about catch it and process it similar to `ri.has_error_registered`?

File: reconcile/terraform_resources.py [hunk 2]
```diff
@@ -415,6 +419,15 @@ def run(
         if err:
             cleanup_and_exit(tf, err)
 
+        if defer:
+            defer(
```
Comment (chassing) [inline reconcile/terraform_resources.py:423]: @maorfr, what do you think?

Comment (maorfr) [reply reconcile/terraform_resources.py:423]: i think you removed `if not light and tf.should_apply:`

Comment (chassing) [reply reconcile/terraform_resources.py:423]: No, take a look at [line 417](https://github.com/app-sre/qontract-reconcile/blob/288ffab049924f1275455a7989b18e78b4348957/reconcile/terraform_resources.py#L417)

**Review by maorfr — APPROVED** [review]

**Pull request merged**